### ISSUE TITLE
PC-98 FDC controller emulation

### ISSUE THREAD

**Issue by joncampbell123** [joncampbell123/dosbox-x#1186]
**Describe the solution you'd like**
Adapt floppy.cpp to operate as the typical NEC-based floppy controller, adapted to the I/O port layout and wiring of the controller in PC-98 mode.

**Additional context**
The floppy controller in PC-98 is the same NEC controller used in IBM PC systems, though on different I/O ports and on a different interrupt, and some I/O ports arrange the bits differently.

Furthermore according to the Undocumented PC-98 reference, there are two floppy controllers, at 90h and the other at C8h, each with their own interrupt (INT41 and INT42).

The FDC at 90h is documented for use with 1MB floppies (1.2MB 3-mode, and 1.44MB on later models), and the one at C8h is documented for use with 640KB (and IBM-compatible 720KB in 1.44MB mode?) floppies.

At least two games in the test collection I have directly talk to the floppy controller after boot: Xanandu-scenario and Dragon Buster. These games will not run in DOSBox-X until FDC emulation is implemented.

**Comment by joncampbell123** [comment]
Question: Does this mean that both controllers talk to the same floppy drives over the same floppy cable? If so, are they both active or are they "bank switched" or enabled one at a time in the same fashion as the IDE controller?

**Comment by joncampbell123** [comment]
I will hold off on this feature until after the August 1st release.

**Comment by joncampbell123** [comment]
Also, despite using the same FDC the on-disk format is different from IBM PC. Instead of using 512 bytes per sector as IBM PC does, PC-98 uses 1024 bytes per sector. The disk RPM differs as well, enough so a 1.2MB 3-mode in an IBM PC produces a data rate out of range of any of the rate selection bits available on IBM's implementation. DOSLIB FDC testing shows that there is one rate that can be used to *almost* read 3-mode, but very unreliable. However if later PC-9821 systems configure their FDC for 3.5" 1.44MB floppies, disks written in one are readable on IBM PC.

**Comment by cracyc** [comment]
Also, the pc98 uses ready polling for disk change detection in the bios so that might have to be implemented if there's software which looks for it.  

> The FDC at 90h is documented for use with 1MB floppies (1.2MB 3-mode, and 1.44MB on later models), and the one at C8h is documented for use with 640KB (and IBM-compatible 720KB in 1.44MB mode?) floppies.

FWIW, I have seen no software which expects the 2DD controller to be available and the bioses in newer machines don't even seem to check for it.

**Comment by joncampbell123** [comment]
@cracyc Then I will implement only the FDC at port 90h, since the two games mentioned are working test cases, and leave a note to implement the second one at C8h if it turns out there is a game that uses it.

**Comment by joncampbell123** [comment]
@cracyc I don't see a ready bit in the I/O port listing but I do see a "busy" bit, do you mean the software polls that bit until it clears?

**Comment by cracyc** [comment]
Nope, from page 5 of https://archive.org/details/bitsavers_necdatashe79_1461697

>After the Specify Command has been sent to the uPD765, the unit select lines USO 
(pin 29) and US1 (pin 28) will automatically go into a scan mode. All four 
FDD's will be scanned by the uPD765 which will be looking for a change in the 
ready line from any of the drives. If the ready line from any of the drives 
should change state (such as due to the door opening or closing) then the uPD765 
will generate an interrupt. When status register (STO) is read (after Sense 
Interrupt Status instruction has been issued) one will find that one of the 
drives will be in the "not ready" mode. This scanning of the ready line by the 
uPD765 occurs continuously between instructions, thus allowing the processor to 
know which drives are "off line".

**Comment by joncampbell123** [comment]
@cracyc According to some others here, there are PC-98 games that do expect an interrupt to occur when disks are changed, so this is important information to keep in mind when implementing. There are at least 4 games in the test suite that I cannot run because they do not see any disk change despite issuing "swap floppy" from the mapper.

**Comment by joncampbell123** [comment]
Very basic FDC emulation is implemented. Unfortunately it appears the reason the two games I wrote it for (Xanadu and Dragon Buster) talk directly to the FDC precisely because they use weird sector formats and out of range C/H/S values (for example, appears to try to read sectors with H=1 from the first side). Dragon Buster appears to assume port 90h and DMA=3 and is not reading the correct data (and crashes).

So perhaps the last part to look at is a disk change interrupt if that helps.

**Comment by cracyc** [comment]
> Dragon Buster appears to assume port 90h and DMA=3 and is not reading the correct data (and crashes).

I just checked in mame and it's only using dma channel 2, 3 isn't touched, so something else must be going wrong here.
